autoretrieve/options.go

```diff
@@ -0,0 +1,273 @@
+	var ras []string
+	for _, ra := range o.provider.Addrs {
+		ras = append(ras, ra.String())
+	}
+	return ras
+}
+
+// WithPurgeCacheOnStart sets whether to clear any cached entries chunks when the provider engine
+// starts.
+// If unset, cache is rehydrated from previously cached entries stored in datastore if present.
+// See: WithDatastore.
+func WithPurgeCacheOnStart(p bool) Option {
+	return func(o *options) error {
+		o.purgeCache = p
+		return nil
+	}
+}
+
+// WithEntriesChunkSize sets the maximum number of multihashes to include in a single entries chunk.
+// If unset, the default size of 16384 is used.
+//
+// See: WithEntriesCacheCapacity, chunker.CachedEntriesChunker
+func WithEntriesChunkSize(s int) Option {
+	return func(o *options) error {
+		o.entChunkSize = s
+		return nil
+	}
+}
+
+// WithEntriesCacheCapacity sets the maximum number of advertisement entries chains to cache.
+// If unset, the default capacity of 1024 is used.
+//
+// The cache is evicted using LRU policy. Note that the capacity dictates the number of complete
+// chains that are cached, not individual entry chunks. This means, the maximum storage used by the
+// cache is a factor of capacity, chunk size and the length of multihashes in each chunk.
+//
+// As an example, for 128-bit long multihashes the cache with default capacity of 1024, and default
+// chunk size of 16384 can grow up to 256MiB when full.
+//
+// See: WithEntriesChunkSize, chunker.CachedEntriesChunker.
+func WithEntriesCacheCapacity(s int) Option {
+	return func(o *options) error {
+		o.entCacheCap = s
+		return nil
+	}
+}
+
+// WithPublisherKind sets the kind of publisher used to announce new advertisements.
+// If unset, advertisements are only stored locally and no announcements are made.
+// See: PublisherKind.
+func WithPublisherKind(k PublisherKind) Option {
+	return func(o *options) error {
+		o.pubKind = k
+		return nil
+	}
+}
+
+// WithHttpPublisherListenAddr sets the net listen address for the HTTP publisher.
+// If unset, the default net listen address of '0.0.0.0:3104' is used.
+//
+// Note that this option only takes effect if the PublisherKind is set to HttpPublisher.
+// See: WithPublisherKind.
+func WithHttpPublisherListenAddr(addr string) Option {
+	return func(o *options) error {
+		o.pubHttpListenAddr = addr
+		return nil
+	}
+}
+
+// WithTopicName sets toe topic name on which pubsub announcements are published.
+// To override the default pubsub configuration, use WithTopic.
+//
+// Note that this option only takes effect if the PublisherKind is set to DataTransferPublisher.
+// See: WithPublisherKind.
+func WithTopicName(t string) Option {
+	return func(o *options) error {
+		o.pubTopicName = t
+		return nil
+	}
+}
+
+// WithTopic sets the pubsub topic on which new advertisements are announced.
+// To use the default pubsub configuration with a specific topic name, use WithTopicName. If both
+// options are specified, WithTopic takes presence.
```

Comment:
>If both options are specified, WithTopic takes presence.
 

looking at `newPublisher` I see that the dt publisher always needs topic name even if pubtopic is provided. The quote above says something different, no?

Reply:
Same as above, https://github.com/filecoin-project/index-provider/

Reply:
That repo is a reference, the code says otherwise, we can address our implementation of the reference repo, and even contribute a fix to it, based on our implementation